desc/singularities.py

```diff
@@ -0,0 +1,733 @@
+        source grid around each singular point. See ``FFTInterpolator`` or
+        ``DFTInterpolator``
+    loop : bool
+        If True, evaluate integral using loops, as opposed to vmap. Slower, but uses
+        less memory.
+
+    Returns
+    -------
+    f : ndarray, shape(eval_grid.num_nodes, kernel.ndim)
+        Integral transform evaluated at eval_grid. Vectors are in rpz basis.
+
+    References
+    ----------
+    .. [1] Malhotra, Dhairya, et al. "Efficient high-order singular quadrature schemes
+       in magnetic fusion." Plasma Physics and Controlled Fusion 62.2 (2019): 024004.
+    .. [2] Malhotra, Dhairya, et al. "Taylor states in stellarators: A fast high-order
+       boundary integral solver." Journal of Computational Physics 397 (2019): 108791.
+
+    """
+    # sanitize inputs, we need everything as jax arrays so they can be indexed
+    # properly in the loops
+    src_data = {key: jnp.asarray(val) for key, val in src_data.items()}
+    eval_data = {key: jnp.asarray(val) for key, val in eval_data.items()}
+
+    if isinstance(kernel, str):
+        kernel = kernels[kernel]
+
+    s, q = interpolator.s, interpolator.q
+
+    out2 = _singular_part(
+        eval_data, eval_grid, src_data, src_grid, s, q, kernel, interpolator, loop
+    )
+    out1 = _nonsingular_part(eval_data, eval_grid, src_data, src_grid, s, kernel, loop)
+    return out1 + out2
+
+
+def _kernel_nr_over_r3(eval_data, src_data, diag=False):
+    # n * r / |r|^3
+    src_x = jnp.atleast_2d(
+        rpz2xyz(jnp.array([src_data["R"], src_data["zeta"], src_data["Z"]]).T)
+    )
+    eval_x = jnp.atleast_2d(
+        rpz2xyz(jnp.array([eval_data["R"], eval_data["zeta"], eval_data["Z"]]).T)
+    )
+    if diag:
+        dx = eval_x - src_x
+    else:
+        dx = eval_x[:, None] - src_x[None]
+    n = rpz2xyz_vec(src_data["e^rho"], phi=src_data["zeta"])
+    n = n / jnp.linalg.norm(n, axis=-1)[:, None]
+    r = safenorm(dx, axis=-1)
+    return safediv(jnp.sum(n * dx, axis=-1), r**3)
+
+
+_kernel_nr_over_r3.ndim = 1
+_kernel_nr_over_r3.keys = ["R", "zeta", "Z", "e^rho"]
+
+
+def _kernel_1_over_r(eval_data, src_data, diag=False):
+    # 1/ |r|
+    src_x = jnp.atleast_2d(
+        rpz2xyz(jnp.array([src_data["R"], src_data["zeta"], src_data["Z"]]).T)
+    )
+    eval_x = jnp.atleast_2d(
+        rpz2xyz(jnp.array([eval_data["R"], eval_data["zeta"], eval_data["Z"]]).T)
+    )
+    if diag:
+        dx = eval_x - src_x
+    else:
+        dx = eval_x[:, None] - src_x[None]
+    r = safenorm(dx, axis=-1)
+    return safediv(1, r)
+
+
+_kernel_1_over_r.ndim = 1
+_kernel_1_over_r.keys = ["R", "zeta", "Z"]
+
+
+def _kernel_biot_savart(eval_data, src_data, diag=False):
+    # K x r / |r|^3
+    src_x = jnp.atleast_2d(
+        rpz2xyz(jnp.array([src_data["R"], src_data["zeta"], src_data["Z"]]).T)
+    )
+    eval_x = jnp.atleast_2d(
+        rpz2xyz(jnp.array([eval_data["R"], eval_data["zeta"], eval_data["Z"]]).T)
+    )
+    if diag:
+        dx = eval_x - src_x
```

Comment:
I get that `diag=False` means it will find the B (for this kernel at least) at each point in eval_data with contributions from every other point in `src_data`...

But what does `diag=True` do? it seems like it just takes everypoint in `eval_data` and then subtracts from the corresponding points (`src_x` and `eval_x` must be same size?)
Sorry, I know from the comments it seems that this should be just evaluating the contribution at one eval point from the local polar grid and not the other points' polar grids, but it is not clear to me how that is enforced with the code, can you clarify it for me?

Reply:
if diag is False, then dx is a NxNx3 array, so coupling all src pts to all eval pts. If diag is True its just Nx3 (the diagonal only), so the polar points around each eval point only contribute to their particular eval pt, not neighbors.

Reply:
so for `diag=True`:
`dx` is N x 3 array
`num = K x dx` is also an `N x 3` array 
`r` before the `if diag` statement is an `(N,) ` array
`r = r[:, None]` makes it an `(N,1)` array again
then the final result is just doing the kernel, evaluating once the K at each node?

This is called in the _singular_part as
```python
k = kernel(
            {key: val for key, val in eval_data.items() if key in keys},
            src_data_polar,
            diag=True,
        ).reshape((eval_grid.num_nodes, kernel.ndim))
```
here src_data_polar is the polar grid points around a single point on the eval_grid? I still don't quite see how `diag` makes this do what we want it to do

Reply:
no, here `src_data_polar` contains one polar point around _each_ evaluation point, and we loop over the different polar offsets `dtheta[i]` and `dzeta[i]`

If we have N eval pts and M polar points around each eval pt, this `src_data_polar` is size N, not size M. We do it this way because we can then vectorize the fft interpolation, since the offsets around each eval pt are the same.

The diag part is basically making sure that polar node i around evaluation point j only contributes to the result at eval pt j, since they're in theory entirely independent calculations, we just vectorize them for speed and efficiency.

Reply:
Ok so this function then will calculate, in one call, the contribution to the integral from one polar point around the intended eval point, and `diag=True` makes sure of that, then we loop over all the offsets to then form the full polar contribution for each eval point. That makes more sense, thanks for the clarification